pkg/storage/tpr/storage_interface.go

```diff
@@ -501,78 +525,82 @@ func (t *store) GuaranteedUpdate(
 			glog.Errorf("checking preconditions (%s)", err)
 			return err
 		}
-		// Create a candidate for the new object by applying the userUpdate func
-		candidate, _, err := userUpdate(curState.obj, *curState.meta)
+		// update the object by applying the userUpdate func & encode it
+		updated, _, err := userUpdate(curState.obj, *curState.meta)
 		if err != nil {
 			glog.Errorf("applying user update: (%s)", err)
 			return err
 		}
-		// Get bytes from the candidate
-		candidateData, err := runtime.Encode(t.codec, candidate)
+		updatedData, err := runtime.Encode(t.codec, updated)
 		if err != nil {
 			glog.Errorf("encoding candidate obj (%s)", err)
 			return err
 		}
-		// If the candidate matches what we already have, then all we need to do is
-		// decode into the out object
-		if bytes.Equal(candidateData, curState.data) {
-			err := decode(t.codec, candidateData, out)
+
+		// figure out what the new "current state" of the object is for this loop iteration
+		var newCurState *objState
+		if bytes.Equal(updatedData, curState.data) {
+			// If the candidate matches what we already have, then all we need to do is
+			// decode into the out object
+			err := decode(t.codec, updatedData, out)
 			if err != nil {
 				glog.Errorf("decoding to output object (%s)", err)
 			}
-			return err
-		}
-		// Otherwise, get an up-to-date copy of the resource we're trying to update
-		// (because it may have changed if we're looping and in a race)
-		newCurObj := t.singularShell("", "")
-		if err := t.Get(ctx, key, "", newCurObj, ignoreNotFound); err != nil {
-			glog.Errorf("getting new current object (%s)", err)
-			return err
+			newCurState = curState
+		} else {
+			// If the candidate doesn't match what we already have, then get an up-to-date copy
+			// of the resource we're trying to update
+			// (because it may have changed if we're looping and in a race)
+			newCurObj := t.singularShell("", "")
```

Comment:
I don't think the arguments to this function are documented anywhere, which would be helpful to have